### PR TITLE
Speeding up PlainResource's (_PlainResourceGroup)

### PR DESCRIPTION
**Pull request by aamalev**
## What do these changes do?

Reduces the complexity of resolving routes.
Resolve route on router with many PlainResource's in some cases O(n) -> O(1)

## Are there changes in behavior for the user?

Improved detection of invalid and unreachable routes.

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

File: aiohttp/web_urldispatcher.py [hunk 1]
```diff
@@ -1052,20 +1113,36 @@ def register_resource(self, resource: AbstractResource) -> None:
                     "already handled by {!r}".format(name, self._named_resources[name])
                 )
             self._named_resources[name] = resource
-        self._resources.append(resource)
+        if append:
+            self._resources.append(resource)
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]: I feel like this would make more sense being split into a separate function (and then not using `self.register_resource()` below), rather than using an awkward keyword argument. Like:

```
def register_resource(...):
    self._register_named_resource(resource)
    self._resources.append(resource)
```

File: aiohttp/web_urldispatcher.py [hunk 2]
```diff
@@ -1052,20 +1113,36 @@ def register_resource(self, resource: AbstractResource) -> None:
...
             raise ValueError("path should be started with / or be empty")
+        # Reuse named resource
+        if name:
+            resource = self._named_resources.get(name)
+            if resource is None:
+                pass
+            elif resource.raw_match(path):
+                return cast(Resource, resource)
+            else:
+                raise ValueError(f"Duplicate {name!r}, already handled by {resource!r}")
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]: This looks like a separate change, it doesn't seem related to the rest of the PR. It also appears to make the below section (line 1142) redundant.

File: aiohttp/web_urldispatcher.py [hunk 3]
```diff
@@ -1052,20 +1113,36 @@ def register_resource(self, resource: AbstractResource) -> None:
...
+        if not ("{" in path or "}" in path or ROUTE_RE.search(path)):
+            if self._resources and isinstance(self._resources[-1], GroupPlainResource):
+                grp = self._resources[-1]
+            else:
+                grp = GroupPlainResource()
+                self.register_resource(grp)
+            resource = grp.add_resource(_requote_path(path), name=name)
+            if name:
+                self.register_resource(resource, append=False)
+            return resource
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]: It's not obvious to me why this has moved before the below block of code?

File: aiohttp/web_urldispatcher.py [hunk 4]
```diff
@@ -414,6 +414,65 @@ def __repr__(self) -> str:
...
+            return None
+
+    def raw_match(self, path: str) -> bool:
+        return path in self._resources
+
+    def get_info(self) -> _InfoDict:
+        return {}
+
+    def url_for(self) -> URL:  # type: ignore[override]
+        raise RuntimeError(".url_for() is not allowed for GroupPlainResource")
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]: Could this cause issues elsewhere?

File: aiohttp/web_urldispatcher.py [hunk 5]
```diff
@@ -414,6 +414,65 @@ def __repr__(self) -> str:
...
+
+    def __repr__(self) -> str:
+        return f"<GroupPlainResource count={len(self._resources)}>"
+
+    def add_resource(self, path: str, name: Optional[str] = None) -> Resource:
+        result = self._resources.get(path)
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py:449]: Maybe just use `.setdefault()` and make this entire thing 1 line?

File: aiohttp/web_urldispatcher.py [hunk 6]
```diff
@@ -414,6 +414,65 @@ def __repr__(self) -> str:
...
+        if path in self._resources:
+            return {}
+        else:
+            return None
+
+    def raw_match(self, path: str) -> bool:
+        return path in self._resources
+
+    def get_info(self) -> _InfoDict:
+        return {}
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]: Could this cause issues as well? The typing information doesn't say that any of the keys are required, so maybe it's fine...

File: aiohttp/web_urldispatcher.py [hunk 7]
```diff
@@ -414,6 +414,65 @@ def __repr__(self) -> str:
...
+        if resource is None:
+            return None, allowed_methods
+
+        return await resource.resolve(request)
+
+    def _match(self, path: str) -> Optional[Dict[str, str]]:
+        if path in self._resources:
+            return {}
+        else:
+            return None
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]:
```suggestion
        return {} if path in self._resources else None
```

Could make the same change to `PlainResource._match()` while we're here.

File: aiohttp/web_urldispatcher.py [hunk 8]
```diff
@@ -414,6 +414,65 @@ def __repr__(self) -> str:
...
+        assert not prefix.endswith("/")
+        assert len(prefix) > 1
+        old = self._resources.copy()
+        self._resources.clear()
+        for path, resource in old.items():
+            resource.add_prefix(prefix)
+            self._resources[prefix + path] = resource
+
+    async def resolve(self, request: Request) -> _Resolve:
+        allowed_methods: Set[str] = set()
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]: This isn't used.

File: aiohttp/web_urldispatcher.py [hunk 9]
```diff
@@ -414,6 +414,65 @@ def __repr__(self) -> str:
...
+    @property
+    def canonical(self) -> str:
+        return "/"
+
+    def freeze(self) -> None:
+        self._routes.clear()
+        for resource in self._resources.values():
+            self._routes.extend(cast(Iterator[ResourceRoute], resource))
+
+    def add_prefix(self, prefix: str) -> None:
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py:431]: Given that the code below returns the inner resource, and not the group, when would this method ever get used?

File: aiohttp/web_urldispatcher.py [hunk 10]
```diff
@@ -414,6 +414,65 @@ def __repr__(self) -> str:
...
+        self._resources: Dict[str, Resource] = {}
+
+    @property
+    def canonical(self) -> str:
+        return "/"
+
+    def freeze(self) -> None:
+        self._routes.clear()
+        for resource in self._resources.values():
+            self._routes.extend(cast(Iterator[ResourceRoute], resource))
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]: Why is the `cast()` needed? The definition just above says this should be a `Resource`, so how is it `Iterator[ResourceRoute]`?

File: aiohttp/web_urldispatcher.py (hunk 9 above)
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py:431]: What happens when calling `add_prefix()` on the `PlainResource` in a group? How is the group going to update its keys?

File: aiohttp/web_urldispatcher.py [hunk 11]
```diff
@@ -414,6 +414,65 @@ def __repr__(self) -> str:
...
 
 
+class GroupPlainResource(Resource):
+    def __init__(self) -> None:
+        super().__init__()
+        self._resources: Dict[str, Resource] = {}
+
+    @property
+    def canonical(self) -> str:
+        return "/"
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]: Is this useful... Are there cases where this might cause problems?

File: aiohttp/web_urldispatcher.py [hunk 12]
```diff
@@ -414,6 +414,65 @@ def __repr__(self) -> str:
         return f"<PlainResource {name} {self._path}>"
 
 
+class GroupPlainResource(Resource):
+    def __init__(self) -> None:
+        super().__init__()
+        self._resources: Dict[str, Resource] = {}
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]:
```suggestion
        self._resources: Dict[str, PlainResource] = {}
```

**Review by Dreamsorcerer — COMMENTED** [review]
Would have been useful to add a description (or fill out anything in the PR template)...

Looks like this could provide an improvement in speed, but only when many plain resources are added together in sequence. There also appear to be a number of edge cases which I'm not sure will work safely with these changes.

File: aiohttp/web_urldispatcher.py (hunk 3 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: To catch the last GroupPlainResource

File: aiohttp/web_urldispatcher.py (hunk 10 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: https://github.com/aio-libs/aiohttp/runs/4636862493?check_suite_focus=true

```
Run make mypy
mypy
aiohttp/web_urldispatcher.py:429: error: Argument 1 to "iter" has incompatible type "Resource"; expected "Iterable[ResourceRoute]"  [arg-type]
aiohttp/web_urldispatcher.py:429: note: Following member(s) of "Resource" have conflicts:
aiohttp/web_urldispatcher.py:429: note:     Expected:
aiohttp/web_urldispatcher.py:429: note:         def __iter__(self) -> Iterator[ResourceRoute]
aiohttp/web_urldispatcher.py:429: note:     Got:
aiohttp/web_urldispatcher.py:429: note:         def __iter__(self) -> Iterator[AbstractRoute]
Found 1 error in 1 file (checked 124 source files)
make: *** [Makefile:89: mypy] Error 1
Error: Process completed with exit code 2.
```

Comment (Dreamsorcerer) [reply aiohttp/web_urldispatcher.py]: Fixed: #6460

File: aiohttp/web_urldispatcher.py [hunk 13]
```diff
@@ -414,6 +414,69 @@ def __repr__(self) -> str:
         return f"<PlainResource {name} {self._path}>"
 
 
+class GroupPlainResource(Resource):
+    def __init__(self) -> None:
+        super().__init__()
+        self._resources: Dict[str, PlainResource] = {}
+
+    @property
+    def resources(self) -> List[AbstractResource]:
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]:
```suggestion
    def resources(self) -> List[PlainResource]:
```

File: aiohttp/web_urldispatcher.py (hunk 11 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: fixed

File: aiohttp/web_urldispatcher.py (hunk 9 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py:431]: Method modifies inner resources and returns None. Used in router.add_prefix

Comment (aamalev) [reply aiohttp/web_urldispatcher.py:431]: The method will be called along the chain  router.add_prefix => grp.add_prefix => plain.add_prefix

File: aiohttp/web_urldispatcher.py (hunk 4 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: The group is almost completely encapsulated in a router

File: aiohttp/web_urldispatcher.py (hunk 2 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: In this section reuse all resources by name.
Next section for new PlainResource or reuse without name PlainResource.
And next section for another new or reuse without name last resource.
Finally section create DynamicResource

File: aiohttp/web_urldispatcher.py (hunk 5 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py:449]: `.setdefault()` requires creation PlainResource. it's expensive

File: aiohttp/web_urldispatcher.py (hunk 6 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: I have no other idea

File: aiohttp/web_urldispatcher.py (hunk 7 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: then the readability of the code will decrease

File: aiohttp/web_urldispatcher.py (hunk 2 above)
Comment (Dreamsorcerer) [reply aiohttp/web_urldispatcher.py]: So, 3rd section does `if resource.name == name and resource.raw_match(path):`. But, if it has a name, wouldn't it already exist in `self._named_resources` and therefore already have done this check above? I'm struggling to see a situation where this 3rd check will ever run and return true.

File: tests/test_client_functional.py [hunk 14]
```diff
@@ -876,8 +876,8 @@ async def redirect(request):
         raise web.HTTPFound(location="/")
 
     app = web.Application()
-    app.router.add_get("/", handler)
-    app.router.add_get("/redirect", redirect)
+    app.router.add_get("/", handler, allow_head=False)
+    app.router.add_get("/redirect", redirect, allow_head=False)
```
Comment (Dreamsorcerer) [inline tests/test_client_functional.py:880]: Is this now required? That potentially makes it a backwards-incompatible change. Although, it looks weird that it was previously using `add_head()` separately anyway...

**Review by asvetlov — CHANGES_REQUESTED** [review]
The PR *might* add a performance boost in some cases (I'm still curious about numbers though).
I support the idea if the performance increase is visible but would see a different implementation.

`UrlDispatcher._resources` could be modified from `List[AbstractResource]` to `List[Union[AbstractResource, _PlainResourceGroup]]`.

The `_PlainResourceGroup` is a hidden container for `PlainResource` mapping that is never exposed to end user. It can have `_PlainResourceGroup.resolve()` and `_PlainResourceGroup.freeze()` methods but have no other `AbstractResource`'s methods and properties.

`UrlDispatcher.resources`, `UrlDispatcher.routes`, `UrlDispatcher.register_resource`, and `UrlDispatcher.add_resource` can be modified to work with `_PlainResourceGroup` directly if needed.

What do you think?

File: aiohttp/web_urldispatcher.py [hunk 15]
```diff
@@ -1028,44 +1099,65 @@ def register_resource(self, resource: AbstractResource) -> None:
         if self.frozen:
             raise RuntimeError("Cannot register a resource into frozen router.")
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]: These checks can be skipped if adding a resource, when there is already a group present. In that case, it only runs `_register_named_resource()`.

Maybe it would be better to put these checks in `_register_named_resource()` and call it unconditionally? Changing to `if not name: return` within that function.

File: aiohttp/web_urldispatcher.py [hunk 16]
```diff
@@ -1028,44 +1099,65 @@ def register_resource(self, resource: AbstractResource) -> None:
...
+                raise ValueError(f"Duplicate {name!r}, already handled by {resource!r}")
+        if not ("{" in path or "}" in path or ROUTE_RE.search(path)):
+            if self._resources and isinstance(self._resources[-1], GroupPlainResource):
+                grp = self._resources[-1]
+            else:
+                grp = GroupPlainResource()
+                self.register_resource(grp)
+            resource = grp.add_resource(_requote_path(path), name=name)
+            if name:
+                self._register_named_resource(resource)
```
Comment (Dreamsorcerer) [inline aiohttp/web_urldispatcher.py]: I think there's an edge case here where the `frozen` check doesn't happen. I'd like to see a test for that, I think something like:
```
add_resource("/foo", "foo")
freeze()
add_resource("/bar", "bar")
```

I suspect a simple solution to then get the test passing would be to ensure the group is frozen. It looks like `grp.add_resource()` currently will allow adding new resources after `grp.freeze()`, which seems incorrect. It freezes the resources, but not itself.

File: aiohttp/web_urldispatcher.py (hunk 7 above)
Comment (Dreamsorcerer) [reply aiohttp/web_urldispatcher.py]: Personally, I find the latter more readable, it's faster for me to see that the function returns correctly in all situations. Plus these simple 1-word literals are exactly the case these inline `if` expressions were made for.

File: aiohttp/web_urldispatcher.py (hunk 9 above)
Comment (Dreamsorcerer) [reply aiohttp/web_urldispatcher.py:431]: Problem is that `.add_resource()` appears to be part of the public API: https://docs.aiohttp.org/en/stable/web_reference.html#aiohttp.web.UrlDispatcher.add_resource

So, if a user calls it, then it returns the `PlainResource`. The `.add_prefx()` method is not listed in the documentation, so not sure if it is considered part of the public API or not: https://docs.aiohttp.org/en/stable/web_reference.html#aiohttp.web.PlainResource

Probably a question for @asvetlov.

**Comment by Dreamsorcerer** [comment]
> The PR _might_ add a performance boost in some cases (I'm still curious about numbers though). I support the idea if the performance increase is visible but would see a different implementation.
> 
> `UrlDispatcher._resources` could be modified from `List[AbstractResource]` to `List[Union[AbstractResource, _PlainResourceGroup]]`.
> 
> The `_PlainResourceGroup` is a hidden container for `PlainResource` mapping that is never exposed to end user. It can have `_PlainResourceGroup.resolve()` and `_PlainResourceGroup.freeze()` methods but have no other `AbstractResource`'s methods and properties.
> 
> `UrlDispatcher.resources`, `UrlDispatcher.routes`, `UrlDispatcher.register_resource`, and `UrlDispatcher.add_resource` can be modified to work with `_PlainResourceGroup` directly if needed.
> 
> What do you think?

I think that might resolve some of the more odd quirks with the current proposal.

File: aiohttp/web_urldispatcher.py (hunk 2 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: For example DynamicResource without name

**Comment by aamalev** [comment]
> The PR _might_ add a performance boost in some cases (I'm still curious about numbers though). I support the idea if the performance increase is visible but would see a different implementation.
> 
> `UrlDispatcher._resources` could be modified from `List[AbstractResource]` to `List[Union[AbstractResource, _PlainResourceGroup]]`.
> 
> The `_PlainResourceGroup` is a hidden container for `PlainResource` mapping that is never exposed to end user. It can have `_PlainResourceGroup.resolve()` and `_PlainResourceGroup.freeze()` methods but have no other `AbstractResource`'s methods and properties.
> 
> `UrlDispatcher.resources`, `UrlDispatcher.routes`, `UrlDispatcher.register_resource`, and `UrlDispatcher.add_resource` can be modified to work with `_PlainResourceGroup` directly if needed.
> 
> What do you think?

Good. So there will be a clearer border between public and private. But the group will also need _PlainResourceGroup.add_prefix

Bench script:
```
import time
import asyncio

from aiohttp.test_utils import make_mocked_request
from aiohttp.web import UrlDispatcher, HTTPOk


async def handler(request):
    return HTTPOk()


async def main():
    router = UrlDispatcher()
    for i in range(10):
        router.add_get(f'/_{i}', handler)
    router.add_get('/a', handler)

    on_grp = len(router._resources) == 1

    request = make_mocked_request('GET', '/a')

    n = 2000000
    start = time.time_ns()
    for i in range(n):
        await router.resolve(request)

    print('on_grp', on_grp, 'with n', n, 'time', (time.time_ns() - start) / n)


asyncio.run(main())
```

Numbers:
on_grp True with n 2000000 time 3613.75
on_grp False with n 2000000 time 10493.992

File: tests/test_client_functional.py (hunk 14 above)
Comment (aamalev) [reply tests/test_client_functional.py:880]: Now required. But in any case next add_head creates an unreachable resource

File: aiohttp/web_urldispatcher.py (hunk 15 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: Maybe raise ValueError?

File: aiohttp/web_urldispatcher.py (hunk 16 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: Covered by test_frozen_plain_group

File: aiohttp/web_urldispatcher.py (hunk 4 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: lost relevance

File: aiohttp/web_urldispatcher.py (hunk 11 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: lost relevance

File: aiohttp/web_urldispatcher.py (hunk 13 above)
Comment (aamalev) [reply aiohttp/web_urldispatcher.py]: fixed

**Comment by Dreamsorcerer** [comment]
I believe this is no longer relevant with the new routing algorithm which also improves routing for all resources.
